### DISCUSSION TITLE
Diving Deep into react-native-unistyles: Insights, Comparisons, and Unique Edges

### DISCUSSION THREAD

**Discussion by vlanemcev** [jpudysz/react-native-unistyles#152] [General]
Hey @jpudysz,

Hope you're doing awesome! First off, major kudos for putting together react-native-unistyles– it's pretty clear you've poured a lot of skill and coffee into this project. 🚀 It's projects like yours that make the React Native space so exciting to be a part of.

So, here's the deal: I've been diving into React Native styling libraries, trying to figure out what's what, and your library caught my eye. It's got some really cool vibes, and I'm seriously considering giving it a spin for my next project. But, you know how it goes – before I jump in, I wanna get the lay of the land.

I've got a couple of things I'm super curious about, and I figured who better to ask than the creator themselves, right?

- Performance & Efficiency: How's library handling things under the hood? Especially curious about missed provider for the theme and using Unistyles runtime for that purposes. It's really interesting because other libraries rely on this approach with provider to much. 

- Unique Features & Advantages: Got any secret sauce in react-native-unistyles that sets it apart from the crowd? Something you're particularly proud of? 

- Ideal Use Cases: Where does library really shine? Got any scenarios or project types where it's the clear MVP?

- Comparison with other libraries: Taking into account the bullets above - it would be nice to see comprehensive comparison table as, for example, the libraries like tanstack-query has.  In my case I'm also looking at gluestack-ui library and its gluestack-style approach and into tamagui as well. What do you think about that ?

Honestly, just looking to get a better feel for where react-native-unistyles stands in the grand tapestry of React Native styling options. Your insights would be super valuable, not just to me but likely to a bunch of other folks out here making decisions.

Thanks a ton for considering my ramble here. Really looking forward to hearing your thoughts!

Cheers!

**Comment by jpudysz** [comment]
Hey @vlanemcev 👋

Thank you for the warm welcome. I'm happy to answer your questions. 😉

- **Performance & Efficiency** - Unistyles boasts a unique architecture among UI libraries. Other libraries are built in JavaScript with an optional Babel plugin. In contrast, Unistyles code spans across 4 languages: JavaScript, Kotlin, Objective-C, and C++. The last one is the most significant because I leverage C++ bindings to make various aspects dynamic. I also perform some computations there, ensuring everything runs smoothly. There's no need for providers, as JSI can inject everything into `global`, allowing me to access `UnistylesRuntime` from anywhere.
- **Unique Features & Advantages** - Unistyles does not share any events with React Native. This means I'm not relying on React Native hooks for theme changes, orientation changes, etc. Everything is implemented natively, so Unistyles decides when to re-render your stylesheets. Thanks to the absence of context and the use of C++ bindings, Unistyles is incredibly fast. You can check the benchmarks in the documentation. For me, the most unique feature is `UnistylesRuntime` with multiple properties that can't be accessed anywhere else. Installing numerous libraries would be necessary to obtain this metadata, and I'm continuously expanding the APIs.
- **Ideal Use Cases:** - Unistyles is not a UI Kit. Instead, you can build your UI Kit on top of it. This decision was made because I run a Software House, and every project involves custom designs created by a team of designers. I build a small UI Kit from the ground up for every project. Thus, another benefit is that I'm also actively using Unistyles commercially in multiple projects. The ideal use case is a project where you want to take charge of your components and build them from the ground up. You care about performance and want to offer users unique features like accessibility, multiple themes, etc. It's also perfect for monorepos, desktop apps, etc. I'm also providing bindings for macOS and Windows.
- **Comparison with Other Libraries** - As mentioned before, Unistyles is not a UI Kit, so it shouldn't be compared to Gluestack/Tamagui. It's closer to Restyle from Shopify. I'm not offering any components, making Unistyles suitable for more advanced users who know how to build their UI kit. Other libraries are also great; you should pick whatever matches your style/project. Honestly, there is no bad choice.

One additional benefit of Unistyles 2.0 is its stability. Just go to the closed issues and filter them by relevance. I'm putting my heart into it, as I'm also using it everywhere.

I hope that answers your questions. 😊

**Reply by vlanemcev** [reply]
Hi, thanks for such a detailed answer! 😊
The things is much clearer now!
